🚀 3단계 - 질문 삭제하기 리팩터링

withbeth opened the pull request:
잘 부탁드립니다 :)

withbeth committed:
docs(README.md): add TODO

withbeth committed:
feat(Answer): add delete feature

withbeth committed:
feat(Answers): 1급 컬렉션 추가

withbeth committed:
feat(Question): add delete including answers feature

withbeth committed:
refactor: QnaService delete question feature

withbeth committed:
refactor : add @DisplayName

withbeth committed:
docs(README.md): update TODO progress

limwoobin commented on src/main/java/qna/domain/Question.java:
```diff
@@ -43,6 +52,33 @@ public Question(Long id, String title, String contents) {
         this.contents = contents;
     }
 
+    public List<DeleteHistory> delete(User loginUser) throws CannotDeleteException {
+        final List<DeleteHistory> result = new ArrayList<>();
+        result.add(deleteQuestion(loginUser));
+        result.addAll(deleteAnswers(loginUser));
+        return Collections.unmodifiableList(result);
+    }
```
Question을 조회하는 조건을 보면 answer에 대해서는 모든 댓글을 가져오게 되어있네요
만약 이미 삭제된 댓글이 있다면 삭제히스토리에 중복으로 들어가지 않을까요?
성능을 고려해서 개선해보시면 좋을것같습니다!

limwoobin replied:
해당 로직에 대한 테스트케이스도 있으면 좋겠네요!

withbeth replied:
> Question을 조회하는 조건을 보면 answer에 대해서는 모든 댓글을 가져오게 되어있네요 만약 이미 삭제된 댓글이 있다면 삭제히스토리에 중복으로 들어가지 않을까요? 

꼼꼼한 피드백 감사합니다 :) 다만 위 부분에 대한 설명을 조금 더 풀어서 부탁드려도 될까요? 
조금 이해가 가지 않아서 명확하게 인식 후 개선하려고 합니다.

withbeth replied:
안녕하세요 우빈님

이틀 정도 고민해봤는데요, 다음 부분에서 막혀서 질문드려요! 편하실때 답변 주시면 감사하겠습니다 

> 이때 Answer는 상태를 보지않기때문에 만약 이미 삭제된 답변이 있다면 같이 조회가 될것

-> 이 부분의 구현에서 막혀있는 상태인데요. Answer에서 삭제되지 않는 Question만 조회 하는 방법이 있을까요?

제가 접근한 방법은 다음과 같습니다.

1. Question 엔티티 자체에 @Where(clause = "deleted = false") 걸기

-> 삭제된 답변을 찾고 싶은 경우도 있을 것 같아 제외 했습니다.

2. Answer 엔티티의 question field에 @Where(clause = "deleted = false")  걸기
```java
@ManyToOne(fetch = FetchType.LAZY)  
@Where(clause = "deleted = false")  
@JoinColumn(foreignKey = @ForeignKey(name = "fk_answer_to_question"), name = "question_id")  
private Question question;
```

-> 애초에 Answer개체 생성시, 삭제된 답변으로 연관관계설정되면 작동하지 않음.

limwoobin commented on src/main/java/qna/domain/Question.java:
```diff
@@ -54,6 +90,7 @@ public boolean isOwner(User writer) {
 
     public void addAnswer(Answer answer) {
         answer.toQuestion(this);
+        answers.add(answer);
     }
```
연관관계 편의메서드 잘 적용해주셧네요 ㅎㅎ

limwoobin commented on src/main/java/qna/domain/Question.java:
```diff
@@ -43,6 +52,33 @@ public Question(Long id, String title, String contents) {
+        }
+    }
+
+    private List<DeleteHistory> deleteAnswers(User loginUser) throws CannotDeleteException {
+        try {
+            return answers.delete(loginUser);
+        } catch (CannotDeleteException cannotDeleteException) {
+            throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
+        }
+    }
```
answer에서 직접 예외를 던지고있는데 여기서 다시 catch 하신 이유가 있을까요?

withbeth replied:
에러메시지 전환을 위해서 캐치하고 있습니다

withbeth replied:
다른 좋은 방법 있으시면 공유 부탁 드릴게요 :)

limwoobin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -56,6 +58,18 @@ public Answer(Long id, User writer, Question question, String contents) {
+    public DeleteHistory delete(User loginUser) throws CannotDeleteException {
+        validateOwner(loginUser);
+        deleted = true;
+        return new DeleteHistory(ContentType.ANSWER, getId(), getWriter(), LocalDateTime.now());
```
기존 toQuestion toAnswer 와 같이 선언하면 어느 객체에 대한
삭제 히스토리인지 더 명확하게 표현할 수 있을것같은데 어떻게생각하시나요 ㅎㅎ

withbeth replied:
toAnswerDeleteHistory()메서드로 삭제 히스토리 생성 부분을 분리해봤어요 :)

limwoobin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -56,6 +58,18 @@ public Answer(Long id, User writer, Question question, String contents) {
+        }
+    }
+
     public boolean isOwner(User writer) {
         return this.writer.equals(writer);
     }
```
해당 메서드 private 으로 닫을 수 있지않을까요??

withbeth replied:
Predicate method라 public으로 구현했는데, 굳이 외부에서 안쓴다면 private으로 바꿔도 되겠네요 :)

limwoobin commented:
추가로 Answer에 ManyToOne의 default fetch type인 EAGER로 선언해주신분이 눈에 띄는데요 ㅎㅎ
EAGER 선언시에는 AnswerRepository의 `findByQuestionAndDeletedFalse` 와 같은 jpql이 실행되었을때
N+1 쿼리가 나갈 위험이 있어서 보통 저는 EAGER자체를 지양하는 편이긴 합니다.
이 부분에 대해서도 같이 공부해보시면 좋을것 같아 추가로 코멘트 남겨드려요

```
@Test
    void test() {
        final Answer saved = target.save(answer);
        manager.flush();
        manager.clear();

        Answer result = target.findByQuestionAndDeletedFalse(answer.getQuestion()).get(0);
    }
```
현재 테스트코드에는 저장 후 바로 조회를 해서 1차캐시덕분에 발생하진 않지만
실무에선 저장하고 바로 조회하지 않는 경우도 있기때문에 한번 고민해보심 좋을듯합니다!

이해가 안가신다면 dm주셔도 좋습니다 :)

limwoobin reviewed (changes requested):
안녕하세요 종은님
3단계 미션 잘 수행해주셨습니다 ㅎㅎ
도메인단위의 테스트까지 작성해주신 부분이 굉장히 꼼꼼하고 인상적이었습니다

다만 이대로 머지하기엔 아쉬운 부분이 있어 코멘트남겨드렸습니다.
확인하시고 다시 요청부탁드립니다!

withbeth committed:
refactor: apply feedbacks

- 삭제 되지 않은 답변 조회시, 삭제되지 않은 답변 목록을 LAZY+FETCH
  JOIN으로 가져올수 있도록 변경
- Test Class의 Repository 변수명 변경
- 이미 삭제된 질문 또는 답변일 경우의 유스케이스 추가
- 질문과 답변 삭제시, DeleteHistory를, 해당 클래스의 정적 메서드
  호출하여 반환하도록 변경

withbeth commented:
다음과 같은 피드백 반영하였습니다.

- 삭제 되지 않은 답변 조회시, 삭제되지 않은 답변 목록을 LAZY+FETCH
  JOIN으로 가져올수 있도록 변경
- Test Class의 Repository 변수명 변경
- 이미 삭제된 질문 또는 답변일 경우의 유스케이스 추가
- 질문과 답변 삭제시, DeleteHistory를, 해당 클래스의 정적 메서드
  호출하여 반환하도록 변경

limwoobin commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -46,6 +46,14 @@ public DeleteHistory(ContentType contentType, Long contentId, User deletedBy, Lo
+
+    public static DeleteHistory fromAnswer(Answer answer) {
+        return new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now());
+    }
```
정적팩토리 좋네요 ㅎㅎ

limwoobin commented on src/main/java/qna/domain/QuestionRepository.java:
```diff
@@ -1,12 +1,16 @@
+        + "left join fetch q.answers.values "
+        + "where q.id = :id "
+        + "and q.deleted = false"
+    )
```
fetch join 으로 작성해주셨군요! 👍 
쿼리 한번으로 모두 처리할 수 있게되었네요 ㅎㅎ

limwoobin reviewed (approved):
안녕하세요 종은님
드디어 JPA미션이 모두 끝이났네요 ㅎㅎ
미션 진행하면서 빠르게 피드백이 반영되는것을 보며 저도 많이 배웠던 시간이었습니다!
정말 고생하셨어요

저와의 미션은 여기까지이지만 이후 미션을 진행하며 
고민되거나 의견을 나누고싶은 부분이 있다면 언제든 편하게 연락주셔도됩니다 ㅎㅎ
그럼 다음미션도 화이팅입니다!
